networkx/algorithms/shortest_paths/weighted.py

```diff
@@ -1044,21 +1060,22 @@ def johnson(G, weight='weight', new_weight=None):
     if not nx.is_weighted(G, weight=weight):
         raise nx.NetworkXError('Graph is not weighted.')
 
-    new_node = nx.utils.generate_unique_node()
-    G.add_weighted_edges_from((new_node, node, 0) for node in G.nodes())
+    dist = {}
+    pred = {}
+    for node in G:
+        dist[node] = 0
+        pred[node] = None
 
     # Calculate distance of shortest paths
-    dist = nx.bellman_ford(G, source=new_node, weight=weight)[1]
-
+    dist = _bellman_ford_relaxation(G, pred, dist, G.nodes(), weight)[1]
     delete = False
     if new_weight is None:
         delete = True
-        new_weight = uuid.uuid1()
+        new_weight = str(uuid.uuid1())
 
     for u, v, w in G.edges(data=True):
         w[new_weight] = w[weight] + dist[u] - dist[v]
```

Comment:
I believed that the point was not to add a new node to the graph. So, one thought is to keep new weights in a dictionary and then pass it as input of ``_djikstra_relaxation`` function.